Install PodDistruptionBudgets when replica greater than 1

rgarcia89 opened the pull request:
Currently there is no pdb deployed for components were the replica is defined greater than 1. This can lead to an outage of the components during a cluster update. The added pdb definitions will be added automatically once the replica size is greater than 1 and ensure that at least one replica is ready before the pod will be drained.

Commit:
Allow user to set topologySpreadConstraints

Co-authored-by: Stephan Mauermann <stephan.mauermann@rtl.com>
Co-authored-by: Stephan Mauermann <stephan.mauermann@rtl.de>
Signed-off-by: Diogo Guerra <diogo.filipe.tomas.guerra@cern.ch>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

geneliu committed:
feat: add configmap checksum annotation for portal

Signed-off-by: Gene Liu <geneliu@users.noreply.github.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
feat: add JOBSERVICE_WEBHOOK_JOB_MAX_RETRY and JOBSERVICE_WEBHOOK_JOB_HTTP_CLIENT_TIMEOUT

Signed-off-by: Shengwen Yu <yshengwen@vmware.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
Update logo path to main for new releases

Signed-off-by: OrlinVasilev <ovasilev@vmware.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

jouve committed:
extraEnvVars for all workloads

Signed-off-by: Cyril Jouve <jv.cyril@gmail.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
feat: add config of reaper.max_update_hours & reaper.max_dangling_hours

Signed-off-by: Shengwen Yu <yshengwen@vmware.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

boniek83 committed:
feat: add podLabels

Signed-off-by: Rafał Boniecki <rafal@boniecki.cc>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

kariya-mitsuru committed:
Add TLS 1.3 support for harbor-nginx

Signed-off-by: Mitsuru Kariya <mitsuru.kariya@nttdata.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
fix: add github-ghcr to the list of PERMITTED_REGISTRY_TYPES_FOR_PROXY_CACHE

Signed-off-by: Shengwen Yu <yshengwen@vmware.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
feat: remove notary

Signed-off-by: Shengwen Yu <yshengwen@vmware.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

kariya-mitsuru committed:
Add TLS 1.3 support for harbor-portal

Signed-off-by: Mitsuru Kariya <mitsuru.kariya@nttdata.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Vad1mo committed:
Update README.md (#1488)

Based on multiple discussions and questions from the community. https://cloud-native.slack.com/archives/CC1E09J6S/p1682520074518419

It would make sense to update the readme and clarify the use of username password

Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

PhilipNelson5 committed:
Allow `CONFIG_OVERWRITE_JSON` to be set in values file (#1147)

The JSON string set in core.configureUserSettings string
will be added to the core secret and loaded into the
environment variable CONFIG_OVERWRITE_JSON

Signed-off-by: Philip Nelson <philipnelson5@gmail.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

rgarcia89 commented:
@Vad1mo ready for review

Vad1mo commented:
I like it. It would be beneficial to make it optional:

- [ ]  add enable/disable flag, on each component an  
- [ ] Allow to set
Min amount

rgarcia89 commented:
Is setting the min amount really necessary? Currently at least one pod is necessary in order for the component to fulfill its tasks. Therefore I think we should be good with the fix value.

arjungandhi committed:
Fix incorrect comment on values.yaml

the commend above the existing secret field specified the wrong information for the what the key of the secret should be

Signed-off-by: Arjun Gandhi <contact@arjungandhi.com>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

PeopleOfPlay commented:
Thanks for the implementation, it's really cool! 

I quite agree with @Vad1mo and I understand your point of view, but harcoding values is never very good, some users may have use cases that push them to change this number. (If, for example, the harbor has a lot of simultaneous users, perhaps two replicas of a certain service is the bare minimum required to maintain the desired quality of service when operating on Kubernetes nodes.) 

In issue #1181 robinkb proposed a rather elegant solution from the [oauth2-proxy chart](https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/templates/poddisruptionbudget.yaml).

I think it would be better to set the default value to 1 and leave it up to the user to choose what best suits their needs. (ref [helm doc](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-default-function))

It might also be cool to have the same logic with maxUnavailable fields. (ref #1182)

sudermanjr committed:
Fix #1127 - remove capabilities check for prometheus

Signed-off-by: Andy Suderman <andy@suderman.dev>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
Add tls checksum

Signed-off-by: Hein-Jan Vervoorn <heinjan.vervoorn@ns.nl>
Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
feat: add pdb

Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
feat: add pdb

Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

Commit:
fix: set podDisruptionBudget.enabled to false

Signed-off-by: Raul Garcia Sanchez <info@raulgarcia.de>

MarcHenriot commented:
@rgarcia89 I have pushed a version of the above changes in my fork, can we merge my version to yours ?

rgarcia89 commented:
> I forgot to sign my commits …. What do we do ?

Yeah I just signed them one by one. Should be fine now

rgarcia89 commented:
@Vad1mo I think this PR is ready to be merged

rgarcia89 committed:
Merge branch 'master' into master

MarcHenriot commented:
There is typo in all pdb, I will push a fix to your fork.

It misses `matchLabels` in the `selector` section.

MarcHenriot committed:
fix: add matchLabels in spec.selector in all pdb

Signed-off-by: Marc Henriot <marck.henriot@gmail.com>

MarcHenriot committed:
feat: add pdb to exporter

Signed-off-by: Marc Henriot <marck.henriot@gmail.com>

rgarcia89 committed:
Merge branch 'goharbor:master' into master

rgarcia89 committed:
Merge pull request #2 from MarcHenriot/master

Fix pdb configuration and add pdb to exporter

rgarcia89 commented:
@Vad1mo any idea when this can be merged?

OrlinVasilev commented:
@zyyw  can you please review that PR thank you!

rgarcia89 committed:
Merge branch 'goharbor:main' into master

rgarcia89 commented:
@MinerYang @Kajot-dev maybe you can have a look here too

slushysnowman commented:
Would be great to get this in - at the moment we're manually adding it for our Harbor setup

Kajot-dev commented:
That would be great to have! We're adding these besides the helm chart currently

balonik commented:
@zyyw please review and approve this MR. It is pending just single approval.

rgarcia89 committed:
Merge branch 'main' into master

OrlinVasilev commented:
@Vad1mo I think that's good to go

phantasm66 commented:
It looks like the only failing workflow integration test is due to an outdated workflow action (`deprecated version of `actions/upload-artifact: v2``). Can that be addressed and this feature be merged? 🙏

rgarcia89 committed:
Merge branch 'goharbor:main' into master

rgarcia89 commented:
All cleared and ready to be merged

quercus-carsten commented:
Hello, would it be possible to have this beneficial improvement reviewed? I'm not certain, but it appears that @zyyw may not have been active recently. Could someone else possibly assist?

Kajot-dev commented:
All the requirements are met. Can somebody proceed with the merge before we hit 2.13.0?

rgarcia89 committed:
Merge branch 'goharbor:main' into master

OrlinVasilev reviewed (approved):
LGTM, it's about time to merge this one!

rgarcia89 commented:
I am really wondering to see how many people it needs to get something merged. Especially given the fact that there are already 3 approvals :sweat_smile:

bupd reviewed (approved):
lgtm

quercus-carsten commented:
User @zyyw was requested to review, but it appears that the user has been inactive since September 2024.
There have been no contributions since then.

Is there anyone in the Harbor project who can help out?

MinerYang reviewed (changes requested):
Hi @rgarcia89 ,
By setting PDB  minAvailable: 1 for each of the pod will have these impacts:
-  **Would create a direct conflict with resource management practices**. For example Kubernetes cannot voluntarily move or evict any of the pods from their current nodes. 
- And a PDB violation can also slow down the rolling upgrade process if new nodes aren't available to place replacement pods on, leading to increased deployment times
- PDBs are only effective for pod running with more than one replica. **A PDB for a single-replica application with minAvailable: 1 will block all voluntary disruptions, preventing node maintenance.**
  
PDB sacrifices cluster flexibility, efficiency, and maintainability. I personally do not see it as a necessarily configuration for pod and as an overhead operation for most of the Harbor deploy scenario.

So please clarify your using scenario and necessity justification  and better to make it optional.

Best,
Miner

Kajot-dev commented:
> Hi @rgarcia89 , By setting PDB minAvailable: 1 for each of the pod will have these impacts:
> 
> * **Would create a direct conflict with resource management practices**. For example Kubernetes cannot voluntarily move or evict any of the pods from their current nodes.
> * And a PDB violation can also slow down the rolling upgrade process if new nodes aren't available to place replacement pods on, leading to increased deployment times
> * PDBs are only effective for pod running with more than one replica. **A PDB for a single-replica application with minAvailable: 1 will block all voluntary disruptions, preventing node maintenance.**
> 
> PDB sacrifices cluster flexibility, efficiency, and maintainability. I personally do not see it as a necessarily configuration for pod and as an overhead operation for most of the Harbor deploy scenario.
> 
> So please clarify your using scenario and necessity justification and better to make it optional.
> 
> Best, Miner

I must disagree.

The whole point of PDB is to prevent k8s from evicting all your pods at once and causing a downtime for your service.

> For example Kubernetes cannot voluntarily move or evict any of the pods from their current nodes.

That would be the case if allowed disruptions was `0` (eg. minAvailable when replicas is 1) Of course it is possible to create bad configuration, via anti-affinity or nodeSelector, but that's like saying we should not use `nodeSelector` since you can specify an invalid one.

Perhaps pdb could be made more configurable or made opt-in, but these are an essential part of any ha deployment. I've been using `pdb` with Harbor on multiple production instances for a couple of years and these have prevented many outages during maintenance, cluster upgrade, etc.

I hope that we can find a solution that works for everyone.

MinerYang reviewed (approved):
lgtm

ricardojdsilva87 commented:
Hello,
This feature was not added to any release yet, it's only on the main branch, the `1.18.3` tag was released around 1 hour ago.
Could you please check?
I've also noticed that these [2 lines ](https://github.com/goharbor/harbor-helm/blame/main/values.yaml#L688-L689)on the `values.yaml` file, are duplicated, but were here before this change was introduced. Most likely the lines were copy pasted across the different services and the already existing values were not noticed.
Thanks!

afreyermuth98 commented:
agreed with @ricardojdsilva87 , do you have an ETA for the release ?

flaviomoringa commented:
Such a simple thing... please merge this asap...  make sure others don't have to suffer what happened with us because of not having this in place.
